text/3503-frontmatter.md

```diff
@@ -0,0 +1,485 @@
+
+Downsides
+- Difficult to parse without assistance from something like `syn` as we'd need to distinguish what the start of a stream is vs content of a string literal
+- Being new syntax, there would be a lot of details to work out, including
+  - How to delineate and label documents
+  - How to allow escaping to avoid conflicts with content in a documents
+  - Potentially an API for accessing the document from within Rust
+- Unfamiliar, new syntax, unclear how it will work out for newer users
+
+### Alternative 5: Regular Comment
+
+Simple header:
+```rust
+#!/usr/bin/env cargo
+/* Cargo.toml:
+[package]
+edition = "2018"
+*/
+
+fn main() {
+}
+```
+
+HEREDOC:
+```rust
+#!/usr/bin/env cargo
+/* Cargo.TOML >>>
+[package]
+edition = "2018"
+<<<
+*/
+
+fn main() {
+}
+```
+The manifest can be a regular comment with a header.  If we were to support
+multiple types of content (manifest, lockfile), we could either use multiple
+comments or HEREDOC.
+This does not prescribe the exact syntax used or supported comments
+
+Benefits
+
+Downsides
+- Unfamiliar syntax
+- New style of structured comment for the ecosystem to support with potential
+  compatibility issues, likely requiring a new edition
+- Assuming it can't be parsed with `syn` and either we need to write a
+  sufficiently compatible comment parser or pull in a much larger rust parser
+  to extract and update comments.
+  - Like with doc comments, this should map to an attribute and then we'd just start the MVP with that attribute
+
+### Alternative 6: Static-site generator frontmatter
+
+```rust
+#!/usr/bin/env cargo
+---
+[package]
+edition = "2018"
+---
+
+fn main() {
+}
+```
+This is a subset/specialization of YAML presentation streams that mirrors people's experience with static site generators:
+- The first line post-shebang-stripping is 3+ dashes, then capture all content until a matching pair of dashes on a dedicated line.  This would be captured into a `#![frontmatter = ""]`.  `frontmatter` attribute is reserved for crate roots.  The 3+ with matching pair is a "just in case" a TOML multi-line string has that syntax in it)
+- Future evolution: Allow a markdown-like infostring on the frontmatter opening dashes to declare the format with `cargo` being the default
+- Future evolution: Allow `frontmatter` attribute on any module
+
+Benefits
+- Visually/syntactically lightweight
+- Has parallels to ideas outside of Rust, building on external knowledge that might exist
+- Easy for cargo to parse and modify
+- Can be leveraged by buck2, meson, etc in the future
+
+Downsides
+- Too general that people might abuse it
+- We've extended the frontmatter syntax, undoing some of the "familiarity" benefit
+- People are used to YAML going in frontmatter (though some systems allow other syntaxes)
+- Doesn't feel very rust-like
+
+### Alternative 7: Extended Shebang
+
+````rust
+#!/usr/bin/env cargo
+# ```cargo
+# [dependencies]
+# foo = "1.2.3"
+# ```
+
+fn main() {}
+````
```

Comment:
@bstrie (moved here for context / easier to follow)
> I would like to consider alternative 7, the extended shebang. I don't think the backticks and the redundant `cargo` specifier should be necessary, producing this:
> 
> ```
> #!/usr/bin/env cargo
> # [dependencies]
> # foo = "1.2.3"
> 
> fn main() {}
> ```
> 
> I think this looks quite good. It's fewer lines than the proposed syntax, and mirrors both the shebang syntax and the attribute syntax.
> 
> I understand that people might want this to be generalizable/extensible, but this could suffice for now and any discussion about how to generalize the "info" portion can be left for a future discussion. If people think that it's important to make it generalizable right now, then I'd be interested to hear some concrete use cases.
> 
> EDIT: although I suppose an unmentioned downside of this is that `# [dependencies]` might look a bit _too_ much like an attribute.

Reply:
As you mentioned, there is the syntax ambiguity and backticks let us "escape" this block of `#` lines.

Yes, the cargo specifier is redundant to the reader but each use of `cargo` is for a different purpose
- One is for execution
- One is for parsing

I've noted in the Future Possibilities that we could relax the requirement on having `cargo` in the infostring in the future.  I would lean towards defaulting to cargo rather than parsing the shebang because shebang parsing is messy.

We also wanted to leave the door open (slightly) for adding additional frontmatter blocks, like if we decide to embed lockfiles.